### PR TITLE
[Exam] Filtering for exam score statistics

### PR DESCRIPTION
**Pull request by kloessst**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Introduces filters for the exam score page statistics, which enables instructors to get a more reasonable picture of the exam statistics.

### Description

- [x] Use `ChangeDetectionStrategy.OnPush` for the exam scores component to improve the user experience
- [x] Add filters for the calculation of exercise statistics: One filters out all un-submitted exams and the other filters out all exercises with only empty submissions. Filters only apply to the first table and don't affect the export of results or the student results table
- [x] Calculate the average scores and the number of participants for each exercise and exerciseGroup while taking the filters into account. This calculation is done on the client to avoid multiple REST requests
- [x] Refactor and extend parts of the `ExamScoresDTO`. Contains now the total number of participants per exercise and exercise group. Each `ExerciseResult` stores a flag now indicating whether non empty submissions exist for this exercise.
- [x] Remove now obsolete calculations on server-side and improve performance in general
- [x] Improve the integration test 

### Steps for Testing

1. Create an exam with all exercise types, and multiple exercises per exercise group. You can also tryout optional exercise groups.
2. Conduct the exam with multiple users. Do not work on some exercise and do not submit the exam for some users. Note down what you did with which user.
3. Assess the exam
4. Check the exam scores page and play around with filters
-> Check that the statistics are correctly calculated

### Screenshots

newExamScoresPage

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java [hunk 1]
```diff
@@ -19,17 +19,20 @@
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestBody;
 
-import de.tum.in.www1.artemis.domain.Exercise;
-import de.tum.in.www1.artemis.domain.ProgrammingExercise;
-import de.tum.in.www1.artemis.domain.Result;
-import de.tum.in.www1.artemis.domain.User;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import de.tum.in.www1.artemis.domain.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ExamService.java:24]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.domain'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968850)

Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ExamService.java:24]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.domain'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968850)

Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ExamService.java:24]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.domain'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968850)

Comment (kloessst) [reply src/main/java/de/tum/in/www1/artemis/service/ExamService.java]: Cool new feature

**Pull request merged**